File: backend/algorithms/src/main/java/projectvice/algorithms/sorting/InsertionSort.java
package projectvice.algorithms.sorting;

public class InsertionSort {
  public static void sort(int[] array) {
    for (int i = 1; i < array.length; i++) {
      int toInsert = array[i];
      int j = i - 1;
      while (j >= 0 && array[j] > toInsert) {
        array[j + 1] = array[j];
        j--;
      }
      array[j + 1] = toInsert;
    }
  }
  
  public static void sort_v2(int[] array) {
    for (int i = 1; i < array.length; i++) {
      int j = 0;
      // find insert position
      while (j < i && array[j] < array[i]) j++;
      if (j < i) {
        int next = array[i];
        while (j <= i) {
          int temp = array[j];
          array[j++] = next;
          next = temp;
        }
      }
    }
  }
}

File: UI/examples/algorithms_problems_typescript/src/sorting/insertionSort.ts
export function insertionSort(array: Array<number>): Array<number> {
  for (let i = 1; i < array.length; i++) {
    let j = i - 1;
    let val = array[i];
    while (j >= 0 && array[j] > val) {
      array[j + 1] = array[j];
      j--;
    };
    array[j + 1] = val;
  }
  return array;
}
